controllers/storagecluster/provider_server.go

```diff
@@ -127,20 +133,23 @@ func (o *ocsProviderServer) createDeployment(r *StorageClusterReconciler, instan
 }
 
 func (o *ocsProviderServer) createService(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) (reconcile.Result, error) {
+	serviceType := corev1.ServiceTypeClusterIP
+	if instance.Spec.ProviderAPIServerServiceType != nil {
+		serviceType = *instance.Spec.ProviderAPIServerServiceType
+	} else if instance.Spec.HostNetwork {
+		serviceType = corev1.ServiceTypeNodePort
+	}
 
-	if instance.Spec.ProviderAPIServerServiceType != "" {
-		switch instance.Spec.ProviderAPIServerServiceType {
-		case corev1.ServiceTypeClusterIP, corev1.ServiceTypeLoadBalancer, corev1.ServiceTypeNodePort:
-		default:
-			err := fmt.Errorf("providerAPIServer only supports service of type %s, %s and %s",
-				corev1.ServiceTypeNodePort, corev1.ServiceTypeLoadBalancer, corev1.ServiceTypeClusterIP)
-			r.Log.Error(err, "Failed to create/update service, Requested ServiceType is", "ServiceType", instance.Spec.ProviderAPIServerServiceType)
-			return reconcile.Result{}, err
-		}
-
+	switch serviceType {
+	case corev1.ServiceTypeClusterIP, corev1.ServiceTypeLoadBalancer, corev1.ServiceTypeNodePort:
+	default:
+		err := fmt.Errorf("providerAPIServer only supports service of type %s, %s and %s",
+			corev1.ServiceTypeNodePort, corev1.ServiceTypeLoadBalancer, corev1.ServiceTypeClusterIP)
+		r.Log.Error(err, "Failed to create/update service, Requested ServiceType is", "ServiceType", instance.Spec.ProviderAPIServerServiceType)
+		return reconcile.Result{}, err
 	}
 
-	desiredService := GetProviderAPIServerService(instance)
+	desiredService := GetProviderAPIServerService(instance.Namespace, serviceType)
```

Comment:
The logic to decide the spec of the service should reside inside the `GetProviderAPIServerService` func. There is no reason to calc outside then send it in

Reply:
we are reusing the calculated serviceType after this func as well.

Reply:
You can read it from the desired service

Reply:
Closing this as I have also mentioned on a newer review